spatialos-sdk/src/worker/schema/component_update.rs

```diff
@@ -0,0 +1,159 @@
+use crate::worker::{
+    component::{Component, ComponentId, Update},
+    schema::{owned::*, ArrayField, FieldId, Object, SchemaField},
+};
+use spatialos_sdk_sys::worker::*;
+use std::marker::PhantomData;
+
+#[derive(Debug)]
+pub struct ComponentUpdate(PhantomData<*mut Schema_ComponentUpdate>);
+
+impl ComponentUpdate {
+    pub fn new<U: Update>(update: &U) -> Owned<Self> {
+        // Create the underlying `Schema_ComponentUpdate` and wrap it in a smart pointer.
+        let mut result: Owned<Self> =
+            unsafe { Owned::new(Schema_CreateComponentUpdate(U::Component::ID)) };
+
+        // Populate the update object from the update data.
+        let component_update = &mut *result;
+        update.into_update(component_update);
+
+        result
+    }
+
+    pub(crate) unsafe fn from_raw<'a>(raw: *mut Schema_ComponentUpdate) -> &'a Self {
+        &*(raw as *mut _)
+    }
+
+    pub fn component_id(&self) -> ComponentId {
+        unsafe { Schema_GetComponentUpdateComponentId(self.as_ptr()) }
+    }
+
+    pub fn deserialize<U: Update>(&self) -> U {
+        // TODO: Does it make sense to assert here? Or should we allow for runtime recovery
+        // in this case?
+        assert_eq!(
+            self.component_id(),
+            U::Component::ID,
+            "Attempting to deserialize component update into incorrect type"
+        );
+
+        U::from_update(self)
+    }
+
+    pub fn field<T: SchemaField>(&self, field: FieldId) -> Option<T::RustType> {
+        T::get_update(self, field)
+    }
+
+    pub fn add_field<T: SchemaField>(&mut self, field: FieldId, value: &T::RustType) {
+        T::add_update(self, field, value);
+    }
+
+    pub fn field_array<T: ArrayField>(&self, field: FieldId) -> Option<Vec<T::RustType>> {
+        if T::has_update(self, field) {
+            Some(self.fields().field_array::<T>(field))
+        } else {
+            None
+        }
+    }
+
+    pub fn add_field_array<T: ArrayField>(&mut self, field: FieldId, value: &[T::RustType]) {
```

Comment:
Should this signature be something like: 

```rust
pub fn add_field_array<T: ArrayField, U: AsRef<[T::RustType]>(&mut self, field: FieldId, value: &U)
```

so we can be a bit more generous with what we take in?

Reply:
I'm not sure that's necessary in this case. It's like taking `&str` vs `AsRef<str>`. In most cases I lean towards taking `&str` or `&[T]` for simplicity, and only using generics when there's a specific, known use case that would benefit from doing so. That said, it doesn't seem like there's any technical reason why we couldn't use the generic version here, so I'm okay with switching to that if you think it's better :)

Reply:
No strong use case but gives a bit more flexibility to the caller. That being said - if it implements `AsRef` then the caller can just do `foo.as_ref()` at the call site. 🤷‍♂